Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3850,6 +3854,225 @@ auto AirIRGenerator::addCallRef(const TypeDefinition& signature, Vector<Expressi
+                jit.call(params[params.proc().resultCount(params.value()->type())].gpr(), WasmEntryPtrTag);
+            });
+
+            append(isEmbedderBlock, Jump);
+            isEmbedderBlock->setSuccessors(continuation);
+        }
+
+        m_currentBlock = continuation;
+    } else {
+        auto pair = emitCallPatchpoint(m_currentBlock, signature, results, args);
+        auto* patchpoint = pair.first;
+        auto exceptionHandle = pair.second;
+        // We need to clobber the size register since the LLInt always bounds checks
+        if (m_mode == MemoryMode::Signaling || m_info.memory.isShared())
+            patchpoint->clobberLate(RegisterSet { PinnedRegisterInfo::get().boundsCheckingSizeRegister });
+
+        patchpoint->setGenerator([=, this](CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+            AllowMacroScratchRegisterUsage allowScratch(jit);
+            prepareForReturnCall(jit, wasmCallerInfo, wasmCalleeInfo);
+            exceptionHandle.generate(jit, params, this);
+            CCallHelpers::Call call = jit.threadSafePatchableNearTailCall();
+            jit.addLinkTask([unlinkedWasmToWasmCalls, call, functionIndex](LinkBuffer& linkBuffer) {
+                unlinkedWasmToWasmCalls->append({ linkBuffer.locationOfNearCall<WasmEntryPtrTag>(call), functionIndex });
+            });
+        });
+    }
+    return { };
+}
+
+void AirIRGenerator::prepareForReturnCall(CCallHelpers& jit, const CallInformation& callerInfo, const CallInformation& calleeInfo)
+{
+    uint32_t callerStackSize = WTF::roundUpToMultipleOf(stackAlignmentBytes(), callerInfo.headerAndArgumentStackSizeInBytes);
+    uint32_t calleeStackSize = WTF::roundUpToMultipleOf(stackAlignmentBytes(), calleeInfo.headerAndArgumentStackSizeInBytes);
+    uint32_t offsetFromSP = m_code.frameSize() + callerStackSize - calleeStackSize;
+    GPRReg scratch = wasmCallingConvention().prologueScratchGPRs[1];
+
+#if CPU(X86_64)
+    jit.subPtr(MacroAssembler::TrustedImm32(sizeof(Register)), MacroAssembler::stackPointerRegister);
+    jit.move(CCallHelpers::Address(MacroAssembler::framePointerRegister, CallFrame::returnPCOffset()), CCallHelpers::Address(MacroAssembler::stackPointerRegister), scratch);
+#elif CPU(ARM64) || CPU(APPLE_ARMV7S) || CPU(RISCV64)
+    jit.loadPtr(CCallHelpers::Address(MacroAssembler::framePointerRegister, CallFrame::returnPCOffset()), jit.linkRegister);
+#else
+    UNREACHABLE_FOR_PLATFORM();
+#endif
+
+    jit.emitRestore(m_code.calleeSaveRegisterAtOffsetList());
+    jit.loadPtr(CCallHelpers::Address(MacroAssembler::framePointerRegister), MacroAssembler::framePointerRegister);
+
+    uint32_t offset = calleeStackSize + sizeof(Register);
+    while (offset) {
+        offset -= sizeof(Register);
+        jit.move(CCallHelpers::Address(MacroAssembler::stackPointerRegister, offset), CCallHelpers::Address(MacroAssembler::stackPointerRegister, offsetFromSP + offset), scratch);
```

Comment:
We shouldn't move the arguments onto the stack twice. This is redundant. Let's just have something a la `emitReturnCallPatchpoint` instead of `emitCallPatchpoint`. We first move them to where a call would, then the call moves them to the header. Let's just skip that intermediary step, since it's expensive for no good reason.